#!/bin/bash

# Write JSON files for ENCODE ATAC pipeline input
# Assumes that reads are paired-end
#module load python/2.7
#module load gsutil

set -e

#fastq_dir="gs://archanaraja/rnaseq/test/pilot/fastq_raw"
#fastq_dir="gs://***REMOVED***-transfer-stanford/montgomery/rat_pilot/RNA-seq/fastq"
#fastq_dir="gs://my-bucket/rat/rat_pilot_data"
#fastq_dir="gs://archanaraja/rnaseq/test_data/pilot_stanford_NIH"
fastq_dir="gs://archanaraja/rnaseq/test_data/pilot_stanford_NIH/rat_blood"
json_dir="input_json/rat_pilot_NIH/rat_blood/"
mkdir -p ${json_dir}

# path to genome reference
#star_reference="gs://archanaraja/rnaseq/references/rat/STAR_reference/star_2.7.0d_03-20-19/Rnor6_v95_star_index.tar.gz"
star_reference="gs://archanaraja/rnaseq/references/rat/STAR_reference/Rnor6_v95_star_index.tar.gz"
rsem_reference="gs://archanaraja/rnaseq/references/rat/rsem_reference/rn6_rsem_reference.tar.gz"
rn6_refFlat="gs://archanaraja/rnaseq/references/rat/genome/refFlat_rn6_v95.txt"
rn6_gtf="gs://archanaraja/rnaseq/references/rat/genome/Rattus_norvegicus.Rnor_6.0.95.gtf"
bowtie2_globin="gs://archanaraja/rnaseq/references/rat/bowtie2_index/rn_globin.tar.gz"
bowtie2_rrna="gs://archanaraja/rnaseq/references/rat/bowtie2_index/rn_rRNA.tar.gz"
bowtie2_phix="gs://archanaraja/rnaseq/references/rat/bowtie2_index/phix.tar.gz"
#suffix of input fastq files
#SUF_R1=_R1_001.trimmed.fastq.gz
#SUF_R2=_R2_001.trimmed.fastq.gz
SUF_R1=_R1_001.fastq.gz
SUF_R2=_R2_001.fastq.gz
docker="araja7/motrpac_rnaseq:v0.1"
#gsutil ls gs://***REMOVED***-transfer-stanford/montgomery/rat_pilot/RNA-seq/fastq|awk -F "/" '{print $8}'|awk '{print substr($0,1,8)}'|sed -e 's/_$//'|cut -f1|sort|uniq
#for i in $indiv;do echo $fastq_dir/$i"_R1_001.trimmed.fastq.gz" $fastq_dir/$i"_R2_001.trimmed.fastq.gz";done
#below line is modified to parse samples from the pilot run format
indiv=$(gsutil ls ${fastq_dir}/*.fastq.gz|awk -F "/" '{print $8}'|awk '{print substr($0,1,8)}'|sed -e 's/_$//'|cut -f1|sort|uniq)
#indiv=$(gsutil ls ${fastq_dir}/*.fastq.gz|awk -F "/" '{print $6}'|awk '{print substr($0,1,8)}'|sed -e 's/_$//'|cut -f1|sort|uniq)
#indiv=$(gsutil ls ${fastq_dir}/*.fastq.gz|awk -F "/" '{print $7}'|awk '{print substr($0,1,8)}'|sed -e 's/_$//'|cut -f1|sort|uniq)
for i in $indiv; do

    # skip "Undetermined" FASTQs generated by bcl2fastq
    if [ `echo "$i" | grep "Undeterm"` ]; then
        continue
    fi

    # name JSON file from FASTQ sample name 
    json_file=${json_dir}/${i}.json

    echo "{" > ${json_file}
    echo "    \"rnaseq_pipeline.fastqr1\" : \"${fastq_dir}/${i}${SUF_R1}\"," >> ${json_file}
    echo "    \"rnaseq_pipeline.fastqr2\" : \"${fastq_dir}/${i}${SUF_R2}\"," >> ${json_file}
    echo "    \"rnaseq_pipeline.SID\" : \"${i}\"," >> ${json_file}
    echo "    \"rnaseq_pipeline.preTrimFastQC.outdir\" : \"fastqc_raw\"," >> ${json_file}
#    echo >> ${json_file}
    echo "    \"rnaseq_pipeline.index_adapter\" : \"AGATCGGAAGAGC\"," >> ${json_file} # required when using Ensembl reference genome, which has non-standard chromosome names 
    echo "    \"rnaseq_pipeline.univ_adapter\" : \"AGATCGGAAGAGC\"," >> ${json_file}
    echo "    \"rnaseq_pipeline.minimumLength\" : 20," >> ${json_file}
    echo "    \"rnaseq_pipeline.postTrimFastQC.outdir\" : \"fastqc_trim\"," >> ${json_file} 
#    echo >> ${json_file}
    echo "    \"rnaseq_pipeline.star_align.star_index\" : \"${star_reference}\"," >> ${json_file}
    echo "    \"rnaseq_pipeline.rsem_quant.rsem_reference\" : \"${rsem_reference}\"," >> ${json_file}
    echo "    \"rnaseq_pipeline.rnametrics.ref_flat\" : \"${rn6_refFlat}\"," >> ${json_file}
    echo "    \"rnaseq_pipeline.featurecounts.gtf_file\" : \"${rn6_gtf}\"," >> ${json_file}
    echo "    \"rnaseq_pipeline.bowtie2_globin.genome_dir\" : \"rn_globin\"," >> ${json_file}
    echo "    \"rnaseq_pipeline.bowtie2_globin.genome_dir_tar\": \"${bowtie2_globin}\"," >> ${json_file}
    echo "    \"rnaseq_pipeline.bowtie2_rrna.genome_dir\" : \"rn_rRNA\"," >> ${json_file}
    echo "    \"rnaseq_pipeline.bowtie2_rrna.genome_dir_tar\": \"${bowtie2_rrna}\"," >> ${json_file}
    echo "    \"rnaseq_pipeline.bowtie2_phix.genome_dir\" : \"phix\"," >> ${json_file}
    echo "    \"rnaseq_pipeline.bowtie2_phix.genome_dir_tar\": \"${bowtie2_phix}\"," >> ${json_file}

    # allocate compute resources
    echo "    \"rnaseq_pipeline.num_threads\" : 4," >> ${json_file}
    echo "    \"rnaseq_pipeline.num_preempt\" : 0," >> ${json_file}
    echo "    \"rnaseq_pipeline.memory\" : 16000," >> ${json_file}
    echo "    \"rnaseq_pipeline.cpus\" : 4," >> ${json_file}
    echo "    \"rnaseq_pipeline.docker\" : \"${docker}\"," >> ${json_file}
    echo "    \"rnaseq_pipeline.disk_space\" : 100" >> ${json_file}
    echo >> ${json_file}
    echo "}" >> ${json_file}

done
